### ISSUE TITLE
How do I keep the connection open?

### ISSUE THREAD

**Issue by ChrisWorks** [joeherold/angular2-sails#20]
Hi Joe

Thanks again for a great module; I have a questions on how to correctly use it.

I have a lot of different angular services; e.g: UserService, AdministrationService, etc. They all use the SailsService to call Sails.

Basic scenario:
`MyProfileComponent -> UserService.updateEmail() -> SailsService.post()...`

My problem is that I need to call `SailsService.connect()` inside all of my function in all my services, otherwise I get this _"Cannot read property 'get' of undefined"_. The angular services does not have the `ngOnInit()` that might have solved this.

I am a bit of a newbie here, so its highly likely me that is doing something wrong - or not doing something more likely.

How do I avoid calling `SailsService.connect()` at the beginning of all function/method in my services. (I have a lot of services and functions). I believe I am (right now) opening a new socket.io connection for every single call to the backend.

Is there a smart singleton pattern that should adopt?

Should all my services extend your SailsService and maybe then I could call the e.g `UserService.connect()` in the `ngOnInit()` in my `MyProfileComponent`?

**Comment by joeherold** [comment]
The service should act as a singleton via dependency injection. Just inject it in e.g. your root or app component and connect there

**Comment by joeherold** [comment]
Oops, closing was a mistake

**Comment by ChrisWorks** [comment]
Ok,

So I import the "SailsModule" in app.module using `SailsModule.forRoot()` and 
I use the SailsService in my AuthenticationService like this:
```
constructor(private _backendService : SailsService, private _userService : UserService) {   }

login(username: string, password: string) {
return this._backendService.post(this.authenticationUrl, { "identifier" : username, "password" : password })
            .map((response: any) => {
               console.log(response);
                
                let user = response.data;
                
                if (user) {
                    localStorage.setItem('currentUser', JSON.stringify(user));
                    this._userService.announceLoginUpdate(true);
                }
            });
    }
```
but where should I call the "first" connect from?

In the constructor of my AuthenticationService?

or `ngOnInit()` in my app.component? even though my app.component does not use it?

or allow autoConnect in my index.html? (and keep this connection?)

I have attached a screenshot from my console, where I dont manually call SailsService.connect() anywhere - so I kinda expect it not to work.
<img width="808" alt="screenshot 2017-06-21 15 24 49" src="https://user-images.githubusercontent.com/4264444/27386499-5a00fd86-5696-11e7-86b2-764bb6f9d99c.png">

**Comment by joeherold** [comment]
can you show me your app.module.ts

**Comment by ChrisWorks** [comment]
Here it is, quite big...  I still want to refactor stuff around and into lazy loaded modules, but first I just want to get everything working.

```typescript
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule, Http,BaseRequestOptions } from '@angular/http';


import { AppComponent } from './app.component';
import { routing }        from './app.routing';

import { AlertService, AuthenticationService } from './_services/index';
import { NavbarComponent } from './navbar/navbar.component';
import { LoginComponent } from './login/login.component';
import { HomeComponent } from './home/home.component';

import { AuthGuard } from './_guards/index';
import { AlertComponent } from './_directives/index';

import { SailsModule } from "angular2-sails";

import { DeviationService } from "./_services/deviation.service";
import { UserService } from "./_services/user.service";
import { SearchModule } from "./deviation/search.module";
import { AdminModule } from "./administration/admin.module";
import { WorkflowModule } from "./administration/workflow/workflow.module";
import { ChartModule } from "./charts/chart.module";
import { AlertModule } from 'ngx-bootstrap';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { MyProfileComponent } from './my-profile/my-profile.component';
import { BsDropdownModule, PaginationModule, ModalModule, SortableModule , DatepickerModule, AccordionModule} from 'ngx-bootstrap';
import { EqualValidatorDirective } from './_directives/equal-validator.directive';


import {CreateDeviationModule} from "./deviation/createdeviation/createdeviation.module";
import {EditDeviationModule} from "./deviation/editdeviation/editdeviation.module";
import {DashboardModule} from "./dashboard/dashboard.module";

import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import {TranslateHttpLoader} from '@ngx-translate/http-loader';
import { SignupComponent } from './login/signup.component';
import { ForgottenpasswordComponent } from './login/forgottenpassword.component';
import { ResetpasswordComponent } from './login/resetpassword.component';




export function HttpLoaderFactory(http: Http) {
    return new TranslateHttpLoader(http, "/assets/i18n/", ".json");
}

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    SailsModule.forRoot(),
    AdminModule.forRoot(),
    WorkflowModule.forRoot(),
    NgbModule.forRoot(),
    AccordionModule.forRoot(),
    BsDropdownModule.forRoot(),
    PaginationModule.forRoot(),
    ModalModule.forRoot(),
    SortableModule.forRoot(),
    DatepickerModule.forRoot(),
    SearchModule,
    routing,
    CreateDeviationModule,
    EditDeviationModule,
    DashboardModule.forRoot(),
    TranslateModule.forRoot({
        loader: {
            provide: TranslateLoader,
            useFactory: HttpLoaderFactory,
            deps: [Http]
        }
    })
  ],
  declarations: [
    AppComponent,
    NavbarComponent,
    AlertComponent,
    LoginComponent,
    HomeComponent,
    MyProfileComponent,
    HomeComponent,
    EqualValidatorDirective,
    SignupComponent,
    ForgottenpasswordComponent,
    ResetpasswordComponent,

  ],
  providers: [
    AuthGuard,
    AlertService,
    AuthenticationService,
    DeviationService,
    UserService,
    BaseRequestOptions
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }

```
This is my index.html

```html
<!doctype html>
<html>
<head>
  <base href="/">
  <meta charset="utf-8">
  
  <title></title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="icon" type="image/x-icon" href="assets/images/favicon.ico">
</head>
<body>
  <app-root>Loading...</app-root>
<script type="text/javascript" src="/assets/js/vendor/sails.io.js" transports='["websocket"]'></script>
<script type="text/javascript">
  io.sails.url = "http://xyz.com:8095";
  io.sails.autoConnect = true;
  //io.sails.environment = "production";
</script>
</body>
</html>
```

**Comment by ChrisWorks** [comment]
I have created a very basic project to re-produce my issue. I have sent you an email on this. If you can find my mistake I would greatly appreciate it.

**Comment by joeherold** [comment]
I will take a look at it tomorrow

**Comment by joeherold** [comment]
you can not use get/post/... without a connection...
change your app.component.ts to the following and then test again:
```typescript
import {Component, OnInit} from '@angular/core';
import {SailsService} from 'angular2-sails'; // import service in app.component

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit { // implement the OnInit lifecycle hook

  title = 'app works!';
  constructor( private _sailsService: SailsService ) { //inject the service
  }
  ngOnInit() { //execute the ngOnInit lifecycle hook
    this._sailsService.connect('http://127.0.0.1:1338'); //connect the service
  }
}
```

**Comment by joeherold** [comment]
I guess, you have to turn off autoconnect in your assets/js/vendor/sails.io.js @ line 1348 to avoid disconnect and reconnecting
```javascript
autoConnect: false,
```

do not use the autoconnect of the sails.io.js file. you connect in your angular app

**Comment by joeherold** [comment]
the connect call is important, because it assigns the _io object

if an connection is already established, it will close the connection and reconnect.

so connect in your app.component, which is your root component, and the connection will be available in all other injected locations

**Comment by ChrisWorks** [comment]
Hi 

I did just as you said and now it works! :) My problem was that I had added it to the provider list inside the app.component. Not sure why, but I was getting some "service not provided" errors and this ended up removing those errors, but it also "broke" the correct way of providing the service across the application.

After removing all references and just adding it back as you said - it works! :)

Thank you.
